### PR TITLE
bugfix: Filter rs that are not part of the current Deployement

### PR DESCRIPTION
**Pull request by zhengjr9**
### Ⅰ. Describe what this PR does
If the selectors of multiple deployments are the same, subsequent tuner operations will be performed to rs that is not the current deployment, resulting in frequent scale-up and scale -down releases

### CODE REVIEW

**Comment by veophi** [comment]
@zhengjr9  Have you met this problem ever？ 
Theoretically, these miss-owned ReplicaSets will be claimed by the `native` deployment controller, so we don't add this logic.

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -81,7 +81,18 @@ func (dc *DeploymentController) getReplicaSetsForDeployment(ctx context.Context,
 	}
 	// List all ReplicaSets to find those we own but that no longer match our
 	// selector. They will be orphaned by ClaimReplicaSets().
-	return dc.rsLister.ReplicaSets(d.Namespace).List(deploymentSelector)
+	all, err := dc.rsLister.ReplicaSets(d.Namespace).List(deploymentSelector)
```
Comment (veophi) [inline pkg/controller/deployment/deployment_controller.go:84]: plz rename `all` as `allRSs`, and `owned` as `ownedRSs`  :)

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -81,7 +81,18 @@ func (dc *DeploymentController) getReplicaSetsForDeployment(ctx context.Context,
...
+	if err != nil {
+		return nil, fmt.Errorf("list %s/%s rs failed:%v", d.Namespace, d.Name, err)
+	}
+	// select rs owner by current deployment
+	owned := make([]*apps.ReplicaSet, 0)
+	for _, rs := range all {
+		if metav1.IsControlledBy(rs, d) {
```
Comment (veophi) [inline pkg/controller/deployment/deployment_controller.go:91]: plz add `rs.DeletionTimestamp.IsZero()` condition also.  :)

**Comment by veophi** [comment]
The DCO failed. Please `git commit --amend -s` to signoff yourself and force push again.

**Comment by zhengjr9** [comment]
> Reference in new

yes,if i had 2 deploy which they has the same selector in the same namespace,the rollout controller while scale down and up another deploy that i didn't update it.that is the logs when i didn't update this deploy but it have  relate a rollout.

```log
urceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set test-debug-1-6db77f98b6 to 2 from 1
I1225 07:02:37.337774       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-message-converged-service-75776fddbc to 2 from 1
I1225 07:02:37.344387       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-system-atomic-service-b599b5644 to 2 from 1
I1225 07:02:37.349581       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-operations-dev-85884b9f77 to 2 from 1
I1225 07:02:37.356030       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-order-converged-service-7686bc9fd9 to 2 from 1
I1225 07:02:37.367093       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-user-converged-service-689658f675 to 2 from 1
I1225 07:02:37.372413       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418754", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-application-atomic-service-6586746865 to 2 from 1
I1225 07:02:37.622697       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418831", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-test-rppll-695d87b49c to 0 from 3
I1225 07:02:37.630095       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418831", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set iecom-ebi-finance-5c6777f8dd to 2 from 1
I1225 07:02:37.645174       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418831", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-enterprise-atomic-service-66c4b5fb5b to 2 from 1
I1225 07:02:38.089751       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set dev-mbpm-etl-85c7bc8f48 to 0 from 2
I1225 07:02:38.194954       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-t-oss-77cf6bb898 to 2 from 1
I1225 07:02:38.531822       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set bard-agent-94c944776 to 0 from 2
I1225 07:02:38.538863       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-svc-demo-796f44f796 to 3 from 2
I1225 07:02:38.549307       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set apollo-agent-595fbfbdb6 to 2 from 1
I1225 07:02:38.558517       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-sentinel-dashboard-dev-85b966f496 to 2 from 1
I1225 07:02:38.589784       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-eureka-service-dev-02-774d8898b4 to 4 from 1
I1225 07:02:38.596439       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set iecom-ebi-export-v1-0-9b985dddf to 2 from 1
I1225 07:02:38.614378       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-file-server-dev-77995c8cc9 to 2 from 1
I1225 07:02:38.629823       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-service-message-atomic-service-6b74544cf to 2 from 1
I1225 07:02:38.637471       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-base-eureka-devloper-58b859dfd8 to 2 from 1
I1225 07:02:38.643121       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-dev-gw-config-77bf676bc7 to 2 from 1
I1225 07:02:38.654720       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-paas-eureka1-dev-8db8558fc to 2 from 1
I1225 07:02:38.665317       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set deployment-iecom-ebi-etl-749dbbb57 to 2 from 1
I1225 07:02:38.672323       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"test-mtu-2", UID:"093623c8-13e6-44fb-b3ee-1da52d95d0c3", APIVersion:"apps/v1", ResourceVersion:"2864418854", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set iam-data-sync-prd-dev-74dc9646c6 to 2 from 1
```

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -81,7 +81,22 @@ func (dc *DeploymentController) getReplicaSetsForDeployment(ctx context.Context,
...
 	// selector. They will be orphaned by ClaimReplicaSets().
-	return dc.rsLister.ReplicaSets(d.Namespace).List(deploymentSelector)
+	allRSs, err := dc.rsLister.ReplicaSets(d.Namespace).List(deploymentSelector)
+	if err != nil {
+		return nil, fmt.Errorf("list %s/%s rs failed:%v", d.Namespace, d.Name, err)
+	}
+	// select rs owner by current deployment
+	ownedRSs := make([]*apps.ReplicaSet, 0)
+	for _, rs := range allRSs {
+		if rs.DeletionTimestamp.IsZero() {
```
Comment (veophi) [inline pkg/controller/deployment/deployment_controller.go:91]: `if !rs.DeletionTimestamp.IsZero() {`

Comment (zhengjr9) [reply pkg/controller/deployment/deployment_controller.go:91]: sorry,it was a silly mistake

File: pkg/controller/deployment/deployment_controller.go [hunk 4]
```diff
@@ -101,7 +116,7 @@ func (dc *DeploymentController) syncDeployment(ctx context.Context, deployment *
 	if reflect.DeepEqual(d.Spec.Selector, &everything) {
 		dc.eventRecorder.Eventf(d, v1.EventTypeWarning, "SelectingAll", "This deployment is selecting all pods. A non-empty selector is required.")
 		if d.Status.ObservedGeneration < d.Generation {
-			d.Status.ObservedGeneration = d.Generation
+			d.Status.ObservedGene7ration = d.Generation
```
Comment (veophi) [inline pkg/controller/deployment/deployment_controller.go:119]: typo

**Comment by veophi** [comment]
> The DCO failed. Please `git commit --amend -s` to signoff yourself and force push again.

@zhengjr9

**Pull request merged**